Justfile

```
@@ -0,0 +1,84 @@
+# FIXME: some of these tests don't work very well on MacOS at the moment. Should fix
+# them at earliest convenience :)
+# Also in addition to this, the `nextest-integration` test is currently disabled on macos
+with_macos_excludes := if os() == "macos" {
+    "--exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
+} else {
+    ""
+}
+nightly_flags := "--features nightly,test_features"
+bash := if os() == "macos" { "/bin/bash" } else { "/usr/bin/bash" }
+
+# all the tests, as close to CI as possible
+test: test-ci test-extra
+
+# only the tests that are exactly the same as the ones in CI
+test-ci: (nextest "stable") (nextest "nightly") check-style
+
+# tests that are as close to CI as possible, but not exactly the same code
+test-extra: check-lychee
+
+# all cargo tests, TYPE is "stable" or "nightly"
+nextest TYPE: (nextest-unit TYPE) (nextest-integration TYPE)
+
+# cargo unit tests, TYPE is "stable" or "nightly"
+nextest-unit TYPE:
+    cargo nextest run \
+        --locked \
+        --workspace \
+        --exclude integration-tests \
+        --cargo-profile dev-release \
+        --profile ci \
+        {{ with_macos_excludes }} \
+        {{ if TYPE == "nightly" { nightly_flags } \
+           else if TYPE == "stable" { "" } \
+           else { error("TYPE is neighter 'nightly' nor 'stable'") } }}
+
+# cargo integration tests, TYPE is "stable" or "nightly"
+[linux]
+nextest-integration TYPE:
+    cargo nextest run \
+        --locked \
+        --package integration-tests \
+        --cargo-profile dev-release \
+        --profile ci \
+        {{ if TYPE == "nightly" { nightly_flags } \
+           else if TYPE == "stable" { "" } \
+           else { error("TYPE is neither 'nightly' nor 'stable'") } }}
+[macos]
+nextest-integration TYPE:
+    @echo "Nextest integration tests are currently disabled on macos!"
+
+# generate a codecov report for RULE
+codecov RULE:
+    #!{{ bash }}
+    set -euxo pipefail
+    source <(cargo llvm-cov show-env --export-prefix)
+    just {{ RULE }}
```

```suggestion
    just_executable() {{ RULE }}
```

Unfortunate that it is necessary to recursively invoke `just` with a use-case as basic as this though… as it no longer provides the strong guarantees before starting to execute commands.

---

Right; I’m thinking of `just` as basically a way to multiplex bash scripts easily, where it’s not too bad to have that kind of recursion. Basically, it doesn’t seem to support easily wrapping commands around another command, because it’s still thought as "similar to make" I guess, and make wouldn’t support that. It’d maybe be better if it were a generic scripting language with support for multiple commands, but…